core/unix/signal.c

```diff
@@ -4838,12 +4860,28 @@ master_signal_handler_C(byte *xsp)
                 dcontext = NULL;
             }
         }
-        if (dcontext == NULL) {
-            /* Check for a temporarily-native thread we're synch-ing with. */
-            tr = thread_lookup(get_sys_thread_id());
-            if (tr != NULL)
-                dcontext = tr->dcontext;
-        }
+    }
+    if (dcontext == NULL &&
+        /* Check for a temporarily-native thread we're synch-ing with. */
+        (sig == SUSPEND_SIGNAL
+#ifdef X86
+         || (INTERNAL_OPTION(safe_read_tls_init) &&
+             /* Check for whether this is a thread with its invalid sentinel magic set.
+              * In this case, we assume that it is either a thread that is currently
+              * temporarily-native via API like DR_EMIT_GO_NATIVE, or a thread in the
+              * clone window. If the thread currently makes a clone, we will find it by
+              * looking it up. If the thread was just cloned, we will not find it yet.
+              * In both cases, we know by inspection of our own code, that it is safe to
```

Comment:
Kind of confusing as to what "both" refers to: "either ... temp-native ... or ... clone window", or "makes a clone...was just cloned".  I'm not sure what the two prior sentences "If...we will find it..." and "If...we will not find it" are adding: I would just remove them.

Reply:
Done